--- a/configure
+++ b/configure
@@ -35,7 +35,7 @@
   FREEBSD=1
 fi
 
-SO_FLAGS="-shared"
+SO_FLAGS="-Wl,--as-needed -shared"
 MAC_OS=0
 uname | grep Darwin > /dev/null 2> /dev/null
 if [ $? = 0 ]; then
@@ -256,7 +256,7 @@
   BUILD_MODULE='Y'
 fi
 
-OPTFLAGS="-O"
+OPTFLAGS="-g -O2"
 
 if [ -z $CC ]; then
   CC=gcc
@@ -278,7 +278,7 @@
 echo "libdir=\$(DESTDIR)$libdir" >> config.mak
 echo "gcinlibdir=\$(libdir)/gcin" >> config.mak
 echo "includedir=\$(DESTDIR)$includedir" >> config.mak
-echo "LDFLAGS=$GTKLDFLAGS -lX11 -ldl -lm" >> config.mak
+echo "LDFLAGS=-Wl,--as-needed $GTKLDFLAGS -lX11 -ldl -lm" >> config.mak
 echo "GCIN_VERSION=$GCIN_VERSION" >> config.mak
 echo "GCIN_TABLE_DIR=\$(datadir_r)/gcin/table" >> config.mak
 echo "GCIN_OGG_DIR=\$(datadir_r)/gcin/ogg" >> config.mak
